pkg/preparer/podprocess/reporter.go

```diff
@@ -0,0 +1,313 @@
+	}
+
+	// Check that the file is executable by root
+	if info.Mode()&0100 != 0100 {
+		return nil, util.Errorf("%s is not executable: perms were %s", config.EnvironmentExtractorPath, info.Mode())
+	}
+
+	if pollInterval == 0 {
+		pollInterval = DefaultPollInterval
+	}
+
+	return &Reporter{
+		sqliteDBPath:             config.SQLiteDatabasePath,
+		environmentExtractorPath: config.EnvironmentExtractorPath,
+		workspaceDirPath:         config.WorkspaceDirPath,
+		logger:                   logger,
+		podStatusStore:           podStatusStore,
+		pollInterval:             pollInterval,
+	}, nil
+}
+
+// Starts the reporter. Quickly returns an error if a startup issue occurs.
+func (r *Reporter) Run(quitCh <-chan struct{}) error {
+	err := r.initWorkspaceDir()
+	if err != nil {
+		return err
+	}
+
+	db, err := r.openAndMaybeMigrateDatabase()
+	if err != nil {
+		return err
+	}
+
+	go r.publishProcessExits(db, quitCh)
+	return nil
+}
+
+// Initializes the workspace dir:
+// 1) create it if it doesn't exist with perms 0600
+// 2) create the workspace file inside of the directory if it doesn't exist (and write 0 value to it)
+func (r *Reporter) initWorkspaceDir() error {
+	// Create the dir if it doesn't exist
+	dirInfo, err := os.Stat(r.workspaceDirPath)
+	if os.IsNotExist(err) {
+		err = os.Mkdir(r.workspaceDirPath, 0600)
+		if err != nil {
+			return util.Errorf("Could not create workspace dir: %s", err)
+		}
+	} else if err != nil {
+		return util.Errorf("Could not stat workspace directory: %s", err)
+	} else {
+		if !dirInfo.IsDir() {
+			return util.Errorf("Configured worskpace directory %s already exists but is not a directory", r.workspaceDirPath)
+		}
+	}
+
+	workspaceFilePath := r.workspaceFilePath()
+	// Check that the workspace file either:
+	// 1) exists and is readable and writable by root
+	// OR
+	// 2) does not exist and is successfully created.
+	info, err := os.Stat(workspaceFilePath)
+	if os.IsNotExist(err) {
+		file, err := os.Create(workspaceFilePath)
+		if err != nil {
+			return util.Errorf("Could not create workspace file at %s: %s", workspaceFilePath, err)
+		}
+
+		err = file.Chmod(0600)
+		if err != nil {
+			return util.Errorf("Could not chmod workspace file at %s: %s", workspaceFilePath, err)
+		}
+
+		// Initialize the file with an id of 0
+		err = r.writeLastID(0)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return util.Errorf("Error running stat() on %s: %s", workspaceFilePath, err)
+	} else {
+		if info.Mode()&0600 != 0600 {
+			return util.Errorf("Workspace file %s exists but is not readable and writable by root: perms %s", workspaceFilePath, info.Mode())
+		}
+	}
+
+	return nil
+}
+
+func (r *Reporter) openAndMaybeMigrateDatabase() (*sql.DB, error) {
+	db, err := sql.Open("sqlite3", r.sqliteDBPath)
+	if err != nil {
+		return nil, util.Errorf("Could not open sqlite database: %s", err)
+	}
+
+	sqlStmt := `
+	create table if not exists finishes (
+	    id integer not null primary key,
```

Comment:
oh man i think I do need it for when we delete values: ```The new key will be unique over all keys currently in the table, but it might overlap with keys that have been previously deleted from the table. To create keys that are unique over the lifetime of the table, add the AUTOINCREMENT keyword to the INTEGER PRIMARY KEY declaration. ```